### PR TITLE
auditd-based file integrity monitoring

### PR DESCRIPTION
**Pull request by alessandrogario**
I have implemented file integrity monitoring using the auditd service. This can come in handy as inotify can only subscribe for a limited amount of entities, and can’t be easily changed at runtime by osquery without risking invalidation of the existing watchers.

Using auditd for file integrity monitoring offers the following advantages:
- Correlates process id and executable path atomically using the audit event records. This means that you don’t risk resolving the process id to the wrong executable path.
- Supports anything that the open() syscall can receive (special directories, NFS/SMB/FUSE mount points, etc.)
- Logs the actual syscall parameters for later inspection.

In general, our implementation can be easily expanded in the future to support more event publishers, since the audit netlink can now be shared. This should also empower subscribers, allowing them to use a more lightweight filtering logic suitable for the task they needs to perform. In order to make auditd suitable for file integrity monitoring, we overcame several performance issues when receiving and processing events from the audit netlink. Additionally, subscribers are no longer blocking each other and/or the netlink reader when processing the event records.

**How auditd-based file integrity monitoring works**

The new implementation works by tracking process/handle creation and destruction by monitoring the execve/exit/exit_group and open/close/dup syscalls.

The file descriptors are saved inside a handle table (one for each process) in the auditd_file_events subscriber. The tables are then used to solve the file descriptors when the following system calls are executed: read/write/mmap.

Rows are only inserted inside the table when a file changes state. Valid state changes are: Open -> Read -> Write -> Close.

Configuration is not yet supported but I plan on adding it, in a way similar to the existing file_events table.

**Issues found and fixed in the original code**

While I was separating the netlink reader and the AuditAssembler logic, I encountered and fixed the following issues:

* [d075b79] Audit event IDs were not correctly handled; according to the documentation (see the commit message for more information) the primary key is made by both the event time and the event id. This caused the class to receive events that perceived as duplicates, tossing them away using an internal history of the last received events.

* [ec57b07] Fixed syscall filtering. This caused subscribers to receive events for syscalls they did not subscribe for.

**Modified or added components:**

**AuditNetlink (new)**
This class is responsible for reading from the audit netlink, and has been implemented to make it possible to have more than one consumer.

This is needed because the existing AuditAssembler component is not capable of sharing the audit handle and it doesn’t support complex filtering logic (which makes it unsuitable for certain use-cases). Another issue is that it does not allow the consumers to subscribe for syscalls that receive different types (and amount) of records (i.e.: open() and mmap()).

Two threads are used, one for receiving the records, and another one for parsing them. Subscribers are given a handle to request new updates.

Configuration options have been all moved here, as having all settings here is better than implementing conflict-resolution logic to merge all the (possibly) different rules requested by each subscriber.

**AuditFim (new)**
This new publisher receives events from the AuditNetlink class, and uses record terminators (AUDIT_EOE) to isolate and dispatch file system changes to the auditd_file_events subscriber. This should have less overhead than attempting to complete the events manually (as the AuditAssembler class does), and was initially implemented because some of the syscalls I listen for (i.e.: read and write) are called very often.

**AuditEventPublisher (changed)**
This class has been updated to use the new AuditNetlink class. This component was not suitable for my use case (see the previous two points), so I decoupled it from the audit netlink reader.

**auditd_file_events (new)**
This table receives the events from the AuditFim publisher, building a file descriptor table for each running process. This is also where filtering is implemented, in order to only show file state changes (vs showing all syscalls).

### CODE REVIEW

**Comment by alessandrogario** [comment]
Here's a sample configuration:

```
 "auditd_fim": {
    "include": [
      "/root/.ssh/%%",
      "/home/%/.ssh/%%"
    ],

    "exclude": [
      "/path/to/file"
    ],

    "show_accesses": "true"
  }
```

The show_accesses variable can be set to false to only show write events.

**Comment by alessandrogario** [comment]
I forgot to introduce some low-level syscalls (create/mknod/mknodat). The changes are smalls, as they are all handled the same way.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 77c997c90bae746e3fc04e4b6a9e4deb96de88da (Job results: [5408](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5408/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 77c997c90bae746e3fc04e4b6a9e4deb96de88da (Job results: [5409](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5409/)) failed one or more tests (**Linux**).

**Comment by theopolis** [comment]
The way I'm going to test this will be a little disorganized. Trying to move fast and include nitpick/style/approach/direction questions together, just want to apologize before hand. :P

I've created a squashed version of all the commits and will be using that to track changes locally as I learn the new implementation and tweak/performance-monitor on my test hosts.

**Comment by theopolis** [comment]
How does the publisher shutdown?

```
sudo ./build/linux/osquery/osqueryi --nodisable_events --nodisable_audit --audit_allow_config --audit_persist --audit_allow_sockets --verbose

osquery> ^D

I0802 19:52:07.823665  1052 auditdnetlink.cpp:798] Uninstalling the audit rules we have installed
I0802 19:52:07.823678  1052 auditdnetlink.cpp:814] Restoring the default configuration for the audit service
    

^C^C^C^C^C
```

Seems like I need to `kill -9` the shell.

**Comment by theopolis** [comment]
I can't get the `process_events` table to work.

I used to be able to do:
```
sudo ./build/linux/osquery/osqueryi \
  --nodisable_events --nodisable_audit --audit_allow_config --audit_persist
```

From reading the code I'm guessing you now need:
```
 --audit_allow_process_events
```

But that's not working either?

**Comment by theopolis** [comment]
So if I try to initialize:
```
I0802 19:55:03.866938  1273 init.cpp:392] osquery initialized [version=2.6.0-21-g5692f91]
I0802 19:55:03.866977  1273 extensions.cpp:286] Could not autoload extensions: Failed reading: /etc/osquery/extensions.load
I0802 19:55:03.867067  1273 init.cpp:651] Error reading config: config file does not exist: /etc/osquery/osquery.conf
I0802 19:55:03.867076  1273 events.cpp:824] Event publisher failed setup: auditfim: Publisher disabled via configuration
I0802 19:55:03.867087  1275 interface.cpp:326] Extension manager service starting: /home/teddy/.osquery/shell.em
I0802 19:55:03.867091  1273 events.cpp:824] Event publisher failed setup: kernel: Cannot access /dev/osquery
I0802 19:55:03.867100  1273 events.cpp:824] Event publisher failed setup: syslog: Publisher disabled via configuration
I0802 19:55:03.867132  1273 events.cpp:1077] Error registering subscriber: auditd_fim_events: Subscriber disabled via configuration
I0802 19:55:03.870287  1279 events.cpp:745] Starting event publisher run loop: audit
I0802 19:55:03.870312  1281 events.cpp:745] Starting event publisher run loop: udev
I0802 19:55:03.870288  1280 events.cpp:745] Starting event publisher run loop: inotify
```

Are there two audit-related publishers? Ideally there should be just one publisher that publishes to multiple subscribers.

File: osquery/events/linux/auditdfim.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    Creat,
+    Mknod,
+    Mknodat,
+    Write,
+    Mmap,
+    Invalid
+  };
+
+  //
+  // Populated from the AUDIT_SYSCALL event record
```
Comment (theopolis) [inline osquery/events/linux/auditdfim.h]: If you need multi-line comments use the long-form docblock:
```
/**
 * @brief Here is a summary.
 *
 * Some more stuff yay!
 */
```

Only use the `///` short-form for documenting a member or method that only needs a 1-liner.

File: osquery/events/linux/auditdfim.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#pragma once
```
Comment (theopolis) [inline osquery/events/linux/auditdfim.h]: nit, this will need a copywrite comment header.

File: osquery/events/linux/auditdfim.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+#include "osquery/events/linux/auditdnetlink.h"
+#include <osquery/events.h>
```
Comment (theopolis) [inline osquery/events/linux/auditdfim.h]: nit, prefer to order includes:
1. C
2. C/OS-specific
3. C++
4. C++ third party libraries
5. Project global, aka the top-level `./include/osquery` using the `<>` search order.
6. Project local, so `""` search order.

And place a `\n\n` between each group so clang-format does not apply a total alphanumeric ordering.

File: osquery/events/linux/auditdnetlink.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <atomic>
```
Comment (theopolis) [inline osquery/events/linux/auditdnetlink.h]: Same general advice from before, comment header and only include what you use.

File: osquery/events/linux/auditdnetlink.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  /// This queue contains unprocessed events, waiting for a
+  /// AuditNetlink::getEvents() call to finalize them.
+  std::vector<AuditEventRecord> queue;
+
+  /// Queue mutex.
+  std::mutex queue_mutex;
+};
+
+class AuditdNetlink final {
+  /// This is the set of rules we have applied when configuring the service.
```
Comment (theopolis) [inline osquery/events/linux/auditdnetlink.h]: Mind placing the private members at the bottom, and prefixing the list with `private:`?

File: osquery/events/linux/auditdnetlink.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  /// This is also what we need to remove when exiting.
+  std::vector<audit_rule_data> installed_rule_list_;
+
+  /// The syscalls we are listening for
+  std::set<int> monitored_syscall_list_;
+
+  //
+  // Common thread data
```
Comment (theopolis) [inline osquery/events/linux/auditdnetlink.h]: I don't think these stanzas help, do you mind removing them?

File: osquery/events/linux/auditdnetlink.h [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  //
+
+  /// The thread that processes the audit events
+  std::unique_ptr<std::thread> processing_thread_;
+
+ public:
+  AuditdNetlink(const AuditdNetlink&) = delete;
+  AuditdNetlink& operator=(const AuditdNetlink&) = delete;
+
+  static AuditdNetlink& getInstance();
```
Comment (theopolis) [inline osquery/events/linux/auditdnetlink.h]: We've been using `get()` as the singleton static accessor, mind applying that nit? :D

File: osquery/events/linux/auditdnetlink.h [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+struct AuditNetlinkSubscriberContext final {
+  /// This queue contains unprocessed events, waiting for a
+  /// AuditNetlink::getEvents() call to finalize them.
+  std::vector<AuditEventRecord> queue;
+
+  /// Queue mutex.
+  std::mutex queue_mutex;
+};
+
+class AuditdNetlink final {
```
Comment (theopolis) [inline osquery/events/linux/auditdnetlink.h]: Checkout the `boost::noncopyable` usage in the code base.

File: osquery/events/linux/tests/audit_tests.cpp [hunk 9]
```diff
@@ -17,16 +17,17 @@
 #include <osquery/tables.h>
 
 #include "osquery/events/linux/audit.h"
+#include "osquery/events/linux/auditdnetlink.h"
+#include "osquery/events/linux/tests/audit_tests_common.h"
 #include "osquery/tests/test_util.h"
 
+#include <cstdint>
+#include <ctime>
```
Comment (theopolis) [inline osquery/events/linux/tests/audit_tests.cpp]: nit, ordering

File: osquery/events/linux/tests/audit_tests_common.cpp [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#include "osquery/events/linux/tests/audit_tests_common.h"
```
Comment (theopolis) [inline osquery/events/linux/tests/audit_tests_common.cpp]: wut? why use another file?

File: osquery/events/linux/tests/audit_tests_common.h [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#pragma once
```
Comment (theopolis) [inline osquery/events/linux/tests/audit_tests_common.h]: Same, this seems like an artifact.

File: osquery/events/linux/tests/auditd_fim_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,799 @@
...
+
+#include <linux/audit.h>
+
+#include <cstdint>
+#include <cstdio>
+#include <ctime>
+#include <string>
+#include <vector>
+
+const std::vector<std::pair<int, std::string>> complete_event_list = {
```
Comment (theopolis) [inline osquery/events/linux/tests/auditd_fim_tests.cpp]: Maybe this content should be held as data in `./tools/tests/`?

File: osquery/tables/events/linux/auditd_fim_events.cpp [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include "osquery/tables/events/linux/auditd_fim_events.h"
```
Comment (theopolis) [inline osquery/tables/events/linux/auditd_fim_events.cpp]: nit, ordering

File: osquery/tables/events/linux/auditd_fim_events.cpp [hunk 14]
```diff
@@ -0,0 +1,569 @@
...
+    if (translated_path[0] != '/') {
+      std::string translated_cwd;
+
+      if (cwd[0] == '"') {
+        translated_cwd = cwd.substr(1, cwd.size() - 2);
+      } else {
+        translated_cwd = cwd;
+      }
+
+      translated_path = translated_cwd + "/" + translated_path;
```
Comment (theopolis) [inline osquery/tables/events/linux/auditd_fim_events.cpp]: Does `operator+` take a `'/'` char?

File: osquery/tables/events/linux/auditd_fim_events.cpp [hunk 15]
```diff
@@ -0,0 +1,569 @@
...
+    }
+
+    // Normalize the path; we could have used 'canonicalize()' but that
+    // accesses the file system and we don't want that (because it may
+    // spawn other events). It is also possible that by the time we
+    // inspect the files on disk the state that caused this event has
+    // changed.
+    namespace boostfs = boost::filesystem;
```
Comment (theopolis) [inline osquery/tables/events/linux/auditd_fim_events.cpp]: Avoid aliasing within code, prefer at the top, where we commonly use `fs = boost::filesystem`.

File: osquery/tables/events/linux/auditd_fim_events.cpp [hunk 16]
```diff
@@ -0,0 +1,569 @@
...
+    boostfs::path normalized_path(translated_path);
+    normalized_path = normalized_path.normalize();
+
+    return normalized_path.string();
+  };
+
+  emitted_row_list.clear();
+
+  // Process the syscall events we received and emit the necessary rows
+  for (const SyscallEvent& syscall : syscall_event_list) {
```
Comment (theopolis) [inline osquery/tables/events/linux/auditd_fim_events.cpp]: It seems like `syscall` could be shadowing method names in rare cases, mind choosing another name?

File: osquery/tables/events/linux/auditd_fim_events.cpp [hunk 17]
```diff
@@ -0,0 +1,569 @@
...
+    auto syscall_type = syscall.type;
+
+    switch (syscall_type) {
```
Comment (theopolis) [inline osquery/tables/events/linux/auditd_fim_events.cpp]: This loop and switch are far too complex, consider moving some of the content into methods.

**Comment by alessandrogario** [comment]
The process_events table is working fine for me using the following flags:

    osqueryd --audit_persist --disable_audit=false --audit_allow_config=true
        --disable_events=false --audit_allow_process_events=true

The reason that it may not work for you is that I have added a new flag to control whether it is enabled or not the same way socket_events works. The default setting is disabled.

**Comment by alessandrogario** [comment]
There is one AuditNetlink class fetching the events from the audit netlink and two publishers. The AuditNetlink class is generic and supports more than one consumer.

The first publisher (audit.cpp, used by socket_events and process_events) was not suitable for my use case for the following reasons:
- It is not possible to listen for two or more syscalls accepting different (both in type and in quantity) arguments.
- Each subscriber is required to state what kind (and how many) records it needs.
- It does not make use of the AUDIT_EOE record terminator to understand if an event is terminated or not (this should be slower).
- The algorithm that understands if an event is complete or not does not correctly support audit events that issue the same record type more than once.
- It used to discard valid events, due to how audit event IDs were handled (fixed, see https://github.com/facebook/osquery/commit/d075b796fbcac8fe995b180aaa25d158adc840ab)
- It used to leak unrequested syscalls to the subscribers in case audit_allow_config was not enabled (fixed, see https://github.com/facebook/osquery/commit/ec57b079814692507d6dd67bebd38c9cc2c14940).
- The filtering rules mix event types with event components (i.e.: AUDIT_SYSCALL is not only the first record in a multi-record sequence, but also the audit event type. The subsequent records, up until the next AUDIT_EOE, are just child records)

In order to avoid refactoring all the existing tables, a second publisher has been implemented.

The AuditdFim component is a lot simpler, as it doesn't need to specifically keep track of which records the subscriber expects to receive (the syscall number is enough) because it makes use of the AUDIT_EOE terminator. This should also be faster than completing them manually.

**Comment by osqueryer** [comment]
:-1: The commit 5819a4beed76c2675161b96d6b726b8a2fe78b80 (Job results: [5432](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5432/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 5819a4beed76c2675161b96d6b726b8a2fe78b80 (Job results: [5433](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5433/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:no_good: The commit 6a3d20ea57a9177c7945f60292637c5919ace9ec (Job results: [2160](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2160/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 6a3d20ea57a9177c7945f60292637c5919ace9ec (Job results: [5441](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5441/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 6a3d20ea57a9177c7945f60292637c5919ace9ec (Job results: [5442](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5442/)) failed one or more tests (**Linux**).

File: osquery/events/linux/auditdfim.h (hunk 2 above)
Comment (alessandrogario) [inline osquery/events/linux/auditdfim.h]: Added!

**Comment by alessandrogario** [comment]
The issue you were experiencing with the osqueryi shell should now be fixed!

**Comment by osqueryer** [comment]
:-1: The commit 5edd3a3c24d234f545e3d5e9ff0df2ea17145c71 (Job results: [5444](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5444/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 5edd3a3c24d234f545e3d5e9ff0df2ea17145c71 (Job results: [5445](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5445/)) failed one or more tests (**Linux**).

**Comment by theopolis** [comment]
We should work to converge both publishers into one. The goal of the subscribers is similar where the publisher seeks to abstract the event construction form processing and storage.

It looks like there are merge conflicts and test failures, should I review anyway, or is there work in progress?

**Comment by osqueryer** [comment]
:no_good: The commit c38f7f7895051ba9abf65a27b4a51f2658392b27 (Job results: [2190](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2190/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 9df2acfa64faeda902bde4ef65334a3358cf778e (Job results: [2200](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2200/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 3d225434772d1264a1b3a3d5f5bb643d94874e29 (Job results: [2203](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2203/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit e784403bd88acf2c5244172665298044840a1082 (Job results: [2204](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2204/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit e0982bc833ac9caa968d9008a7d3c7c991d4d173 (Job results: [2205](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2205/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 71d457d00dcda61023309c6b6773ca104b209e5a (Job results: [2209](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2209/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 353d5c5d2bad98625c0d44f07a4486f5b3cf9543 (Job results: [2213](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2213/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit b2c377381be29adad9baa90187026cafd33c71e3 (Job results: [2214](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2214/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 02ff2ad2e726d47e1e8e14436f115e6ab4565e54 (Job results: [2240](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2240/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 0d6cec1933373b1f72a37a87992bc13fb2d7cdf3 (Job results: [2241](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2241/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit 426ee883bccfa6675fdcd6741fb9673037277b3e (Job results: [2246](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2246/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit f8f08f1e925a82815ebe5848768d29f5ffcc21da (Job results: [2249](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2249/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:no_good: The commit bbdf93c80f5c1c9f5592f9bc4b71257307506cbd (Job results: [2250](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2250/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit bbdf93c80f5c1c9f5592f9bc4b71257307506cbd (Job results: [1647](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1647/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit 05a36004c0238824fce7b9531ed27fc475dc43d8 (Job results: [2252](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2252/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 05a36004c0238824fce7b9531ed27fc475dc43d8 (Job results: [499](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/499/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 05a36004c0238824fce7b9531ed27fc475dc43d8 (Job results: [5542](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5542/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 05a36004c0238824fce7b9531ed27fc475dc43d8 (Job results: [5543](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5543/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 05a36004c0238824fce7b9531ed27fc475dc43d8 (Job results: [500](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/500/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [1651](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1651/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [501](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/501/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [1652](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1652/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [502](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/502/)) failed one or more tests (**FreeBSD**).

**Comment by obelisk** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [503](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/503/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 2757444ed8dc016d90cd3ffff034dce42964ebbb (Job results: [504](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/504/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit feb09a46d63d8e26c1dbd4e0b6862501f0ca54c4 (Job results: [5546](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5546/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit feb09a46d63d8e26c1dbd4e0b6862501f0ca54c4 (Job results: [1654](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1654/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit feb09a46d63d8e26c1dbd4e0b6862501f0ca54c4 (Job results: [5547](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5547/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit feb09a46d63d8e26c1dbd4e0b6862501f0ca54c4 (Job results: [505](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/505/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit feb09a46d63d8e26c1dbd4e0b6862501f0ca54c4 (Job results: [506](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/506/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 251efd29af3b0e5129d6803ea7803c862a2b4be5 (Job results: [507](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/507/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit 251efd29af3b0e5129d6803ea7803c862a2b4be5 (Job results: [508](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/508/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit fc038508c33f6fed3cb6b92aa3c37858abeb5ffe (Job results: [2454](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2454/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit fc038508c33f6fed3cb6b92aa3c37858abeb5ffe (Job results: [2455](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2455/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit d19eecf42375cc27803e38713d9cc4dfd739564c (Job results: [2456](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2456/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit d19eecf42375cc27803e38713d9cc4dfd739564c (Job results: [2457](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2457/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit bb426e7d1f5c3acb9274deeb82d70138d32b557e (Job results: [2458](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2458/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit bb426e7d1f5c3acb9274deeb82d70138d32b557e (Job results: [2459](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2459/)) failed one or more tests (**Windows**).

**Pull request merged**